api_inference_community/hub.py

```diff
@@ -0,0 +1,165 @@
+    with open(card_path, "r") as f:
+        data = f.read()
+
+    card_chunks = re.split(YAML_DOC_SEP, data)
+
+    card_data = "{}"
+    for c in card_chunks:
+        c = c.strip()
+        if c:
+            card_data = c
+            break
+    card_data = yaml.safe_load(card_data)
+    siblings = _build_offline_siblings(repo)
+    model_info = hf_api.ModelInfo(
+        private=False,
+        downloads=0,
+        likes=0,
+        id=repo_id,
+        card_data=card_data,
+        siblings=siblings,
+        **card_data,
+    )
+
+    logger.info("Offline model info for repo %s: %s", repo, model_info)
+
+    return model_info
+
+
+def _build_offline_siblings(repo: pathlib.Path) -> List[dict]:
+    siblings = []
+    prefix_pattern = re.compile(r"^" + re.escape(str(repo)) + r"(.*)$")
+    for root, dirs, files in os.walk(repo):
+        for file in files:
+            filepath = os.path.join(root, file)
+            size = os.stat(filepath).st_size
+            m = prefix_pattern.match(filepath)
+            if not m:
+                msg = (
+                    f"File {filepath} does not match expected pattern {prefix_pattern}"
+                )
+                logger.error(msg)
+                raise Exception(msg)
+            filepath = m.group(1)
+            filepath = filepath.strip(os.sep)
+            sibling = dict(rfilename=filepath, size=size)
+            siblings.append(sibling)
+    return siblings
+
+
+def _cached_model_info(
+    repo_id: str, revision: str, cache_dir: pathlib.Path
+) -> hf_api.ModelInfo:
+    revision_path = cached_revision_path(cache_dir, repo_id, revision)
+    model_info_basename = "hub_model_info.json"
+    model_info_path = revision_path / model_info_basename
+    logger.info("Checking if there are some cached model info at %s", model_info_path)
+    if os.path.exists(model_info_path):
+        with open(model_info_path, "r") as f:
+            o = json.load(f)
+        r = hf_api.ModelInfo(**o)
+        logger.debug("Cached model info from file: %s", r)
+    else:
+        logger.debug(
+            "No cached model info file %s found, "
+            "rebuilding partial model info from cached model files",
+            model_info_path,
+        )
+        # Let's rebuild some partial model info from what we see in cache, info extracted should be enough
+        # for most use cases
+        r = _build_offline_model_info(repo_id, revision_path)
+
+    return r
+
+
+def hub_model_info(
+    repo_id: str,
+    revision: Optional[str] = None,
+    cache_dir: Optional[pathlib.Path] = None,
+    **kwargs,
+) -> hf_api.ModelInfo:
+    """
+    Get Hub model info with offline support, provided some hub_model_info.json file has been filled in the
+    local model directory
+    """
+    if revision is None:
+        revision = "main"
+
+    if not constants.HF_HUB_OFFLINE:
+        return hf_api.model_info(repo_id=repo_id, revision=revision, **kwargs)
+
+    logger.info("Model info for offline mode")
```

Comment:
Thanks @Wauplin for all the tips! so I updated the pr accordingly
1. Used the try_load_cache + ModelCard.load tip, thanks
2. Could not totally get rid of the cached_revsision_path method though, for several reasons:
  a. if we keep the hub_model_info.json convenience (but this would definitely not be a sufficient reason in itself)
  b. because if we focus on the setfit fix here: https://github.com/huggingface/api-inference-community/pull/378/commits/38c13a4d33e474629ad48475355d1763851c6222, what fixes it is that we replace the hub repo_id with the local path to work around the issues with HF_HUB_OFFLINE in sentence_transformers, not th whole model_info mechanic that we added in prevision of fixing this https://github.com/huggingface/api-inference-community/issues/372
3. Tried to apply huggingface_hub.utils.OfflineModeIsEnabled trick with the latest hf_hub lib (0.20.2) but for some reason my unittests did not work -> I guess it's just because in the tests I override the constasnts.HF_HUB_OFFLINE, but it's too late, but I first need to be sure